src/wasm-binary-reader-objdump.c

```diff
@@ -0,0 +1,434 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "wasm-binary-reader-objdump.h"
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+
+#include "wasm-binary-reader.h"
+
+typedef struct Context {
+  const WasmObjdumpOptions* options;
+  WasmStream* out_stream;
+  const uint8_t* data;
+  size_t size;
+  WasmOpcode current_opcode;
+  size_t current_opcode_offset;
+  size_t last_opcode_end;
+  int indent_level;
+} Context;
+
+static void WASM_PRINTF_FORMAT(2, 3)
+    print_details(Context* ctx, const char* fmt, ...) {
+  if (!ctx->options->verbose)
+    return;
+  va_list args;
+  va_start(args, fmt);
+  vprintf(fmt, args);
+  va_end(args);
+}
+
+#define SEGSTART(segname, name)                                             \
+  static WasmResult begin_##segname##_section(WasmBinaryReaderContext* ctx, \
+                                              uint32_t size) {              \
+    return begin_section(ctx->user_data, name, ctx->offset, size);          \
+  }
+
+static WasmResult begin_section(Context* ctx,
+                                const char* name,
+                                size_t offset,
+                                size_t size) {
+  if (ctx->options->headers) {
+    printf("%-12s start=%#010" PRIzx " end=%#010" PRIzx " (size=%#010" PRIzx
```

Comment:
Actually, maybe start and end, too? IIRC the JS engines report byte offsets in Decimal in error messages stacktraces.